File: core/shared/schemas.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Dict, Any, Optional, List

class NodeInfo(BaseModel):
    node_id: str
    os_name: str
    ip_address: str
    openclaw_version: Optional[str] = None
    hostname: Optional[str] = None

class SystemMetricsPayload(BaseModel):
    node_id: str
    cpu_percent: float
    cpu_count: int = 0
    load_avg_1m: float = 0.0
    load_avg_5m: float = 0.0
    load_avg_15m: float = 0.0
    ram_used_mb: float
    ram_total_mb: float
    ram_percent: float = 0.0
    swap_used_mb: float = 0.0
    swap_total_mb: float = 0.0
    disk_used_percent: float
    disk_total_gb: float = 0.0
    uptime_seconds: int = 0
    boot_time_seconds: int = 0
    net_tx_bytes: int = 0
    net_rx_bytes: int = 0
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class AgentEventPayload(BaseModel):
    node_id: str
    session_id: str
    event_type: str
    content: Dict[str, Any] = Field(default_factory=dict)
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    model: Optional[str] = None
    provider: Optional[str] = None
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    cost_usd: float = 0.0
    duration_ms: int = 0
    tool_name: Optional[str] = None
    channel: Optional[str] = None

class WorkspaceFilePayload(BaseModel):
    path: str
    filename: str
    type: str
    content: Optional[str] = None

class WorkspaceSummary(BaseModel):
    agent_name: str
    soul_exists: bool
    agents_exists: bool
    memory_exists: bool
    state_exists: bool
    soul_content: Optional[str] = None
    agents_content: Optional[str] = None
    state_content: Optional[str] = None
    heartbeat_ms_ago: int = 0
    daily_notes_count: int = 0

class WorkspaceEventPayload(BaseModel):
    node_id: str
    type: str
    files: List[WorkspaceFilePayload] = Field(default_factory=list)
    summary: Optional[WorkspaceSummary] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class CronRunRecord(BaseModel):
    ts: int
    status: str
    summary: Optional[str] = None
    error: Optional[str] = None
    sessionId: Optional[str] = None
    durationMs: Optional[int] = None

class CronJob(BaseModel):
    id: str
    name: str
    schedule: Any
    enabled: bool = True
    payload: Optional[Any] = None
    last_run_ms: int = 0
    next_run_ms: int = 0
    run_count: int = 0
    error_count: int = 0
    recent_runs: Optional[List[CronRunRecord]] = []

class CronEventPayload(BaseModel):
    node_id: str
    type: str
    jobs: List[CronJob] = Field(default_factory=list)
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class SessionEntry(BaseModel):
    sessionId: str
    key: str
    label: Optional[str] = None
    model: Optional[str] = None
    provider: Optional[str] = None
    status: Optional[str] = None
    created_at_ms: int = 0
    last_active_ms: int = 0
    token_count: int = 0
    cost_usd: float = 0.0
    agent_name: Optional[str] = None
    channel: Optional[str] = None
    is_history: bool = False

class SessionsEventPayload(BaseModel):
    node_id: str
    type: str
    sessions: List[SessionEntry] = Field(default_factory=list)
    session_count: int = 0
    active_count: int = 0
    history_count: int = 0
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class GatewayLogEventPayload(BaseModel):
    node_id: str
    type: str
    log_path: str
    lines: List[Dict[str, Any]] = Field(default_factory=list)
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class HealthSnapshot(BaseModel):
    timestamp: int
    cpu_percent: float
    ram_percent: float
    disk_percent: float
    temp_cpu: Optional[float] = None

class HealthHistoryEventPayload(BaseModel):
    node_id: str
    type: str
    snapshots: List[HealthSnapshot] = Field(default_factory=list)
    count: int = 0
    timestamp: datetime = Field(default_factory=datetime.utcnow)

File: services/server/schemas/schemas.go
package schemas

import (
	"time"

	"gorm.io/datatypes"
)

type NodeInfo struct {
	NodeID          string  `json:"node_id"`
	OSName          string  `json:"os_name"`
	IPAddress       string  `json:"ip_address"`
	OpenClawVersion *string `json:"openclaw_version,omitempty"`
	Hostname        *string `json:"hostname,omitempty"`
}

type SystemMetricsPayload struct {
	NodeID          string    `json:"node_id"`
	CPUPercent      float64   `json:"cpu_percent"`
	CPUCount        int       `json:"cpu_count"`
	LoadAvg1m       float64   `json:"load_avg_1m"`
	LoadAvg5m       float64   `json:"load_avg_5m"`
	LoadAvg15m      float64   `json:"load_avg_15m"`
	RAMUsedMB       float64   `json:"ram_used_mb"`
	RAMTotalMB      float64   `json:"ram_total_mb"`
	RAMPercent      float64   `json:"ram_percent"`
	SwapUsedMB      float64   `json:"swap_used_mb"`
	SwapTotalMB     float64   `json:"swap_total_mb"`
	DiskUsedPercent float64   `json:"disk_used_percent"`
	DiskTotalGB     float64   `json:"disk_total_gb"`
	UptimeSeconds   int       `json:"uptime_seconds"`
	BootTimeSeconds int       `json:"boot_time_seconds"`
	NetTxBytes      int64     `json:"net_tx_bytes"`
	NetRxBytes      int64     `json:"net_rx_bytes"`
	Timestamp       time.Time `json:"timestamp"`
}

type AgentEventPayload struct {
	NodeID           string         `json:"node_id"`
	SessionID        string         `json:"session_id"`
	EventType        string         `json:"event_type"`
	Content          datatypes.JSON `json:"content"`
	Timestamp        time.Time      `json:"timestamp"`
	Model            *string        `json:"model,omitempty"`
	Provider         *string        `json:"provider,omitempty"`
	InputTokens      int            `json:"input_tokens"`
	OutputTokens     int            `json:"output_tokens"`
	CacheReadTokens  int            `json:"cache_read_tokens"`
	CacheWriteTokens int            `json:"cache_write_tokens"`
	CostUSD          float64        `json:"cost_usd"`
	DurationMs       int            `json:"duration_ms"`
	ToolName         *string        `json:"tool_name,omitempty"`
	Channel          *string        `json:"channel,omitempty"`
}

type WorkspaceFilePayload struct {
	Path     string  `json:"path"`
	Filename string  `json:"filename"`
	Type     string  `json:"type"`
	Content  *string `json:"content,omitempty"`
}

type WorkspaceSummary struct {
	AgentName       string  `json:"agent_name"`
	SoulExists      bool    `json:"soul_exists"`
	AgentsExists    bool    `json:"agents_exists"`
	MemoryExists    bool    `json:"memory_exists"`
	StateExists     bool    `json:"state_exists"`
	SoulContent     *string `json:"soul_content,omitempty"`
	AgentsContent   *string `json:"agents_content,omitempty"`
	StateContent    *string `json:"state_content,omitempty"`
	HeartbeatMsAgo  int     `json:"heartbeat_ms_ago"`
	DailyNotesCount int     `json:"daily_notes_count"`
}

type WorkspaceEventPayload struct {
	NodeID    string                 `json:"node_id"`
	Type      string                 `json:"type"`
	Files     []WorkspaceFilePayload `json:"files"`
	Summary   *WorkspaceSummary      `json:"summary,omitempty"`
	Timestamp time.Time              `json:"timestamp"`
}

type CronRunRecord struct {
	Ts         int     `json:"ts"`
	Status     string  `json:"status"`
	Summary    *string `json:"summary,omitempty"`
	Error      *string `json:"error,omitempty"`
	SessionId  *string `json:"sessionId,omitempty"`
	DurationMs *int    `json:"durationMs,omitempty"`
}

type CronJob struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	Schedule    interface{}     `json:"schedule"`
	Enabled     bool            `json:"enabled"`
	Payload     interface{}     `json:"payload,omitempty"`
	LastRunMs   int             `json:"last_run_ms"`
	NextRunMs   int             `json:"next_run_ms"`
	RunCount    int             `json:"run_count"`
	ErrorCount  int             `json:"error_count"`
	RecentRuns  []CronRunRecord `json:"recent_runs,omitempty"`
}

type CronEventPayload struct {
	NodeID    string    `json:"node_id"`
	Type      string    `json:"type"`
	Jobs      []CronJob `json:"jobs"`
	Timestamp time.Time `json:"timestamp"`
}

type SessionEntry struct {
	SessionId    string  `json:"sessionId"`
	Key          string  `json:"key"`
	Label        *string `json:"label,omitempty"`
	Model        *string `json:"model,omitempty"`
	Provider     *string `json:"provider,omitempty"`
	Status       *string `json:"status,omitempty"`
	CreatedAtMs  int     `json:"created_at_ms"`
	LastActiveMs int     `json:"last_active_ms"`
	TokenCount   int     `json:"token_count"`
	CostUsd      float64 `json:"cost_usd"`
	AgentName    *string `json:"agent_name,omitempty"`
	Channel      *string `json:"channel,omitempty"`
	IsHistory    bool    `json:"is_history"`
}

type SessionsEventPayload struct {
	NodeID       string         `json:"node_id"`
	Type         string         `json:"type"`
	Sessions     []SessionEntry `json:"sessions"`
	SessionCount int            `json:"session_count"`
	ActiveCount  int            `json:"active_count"`
	HistoryCount int            `json:"history_count"`
	Timestamp    time.Time      `json:"timestamp"`
}

type GatewayLogEventPayload struct {
	NodeID    string                   `json:"node_id"`
	Type      string                   `json:"type"`
	LogPath   string                   `json:"log_path"`
	Lines     []map[string]interface{} `json:"lines"`
	Timestamp time.Time                `json:"timestamp"`
}

type HealthSnapshot struct {
	Timestamp   int      `json:"timestamp"`
	CPUPercent  float64  `json:"cpu_percent"`
	RAMPercent  float64  `json:"ram_percent"`
	DiskPercent float64  `json:"disk_percent"`
	TempCPU     *float64 `json:"temp_cpu,omitempty"`
}

type HealthHistoryEventPayload struct {
	NodeID    string           `json:"node_id"`
	Type      string           `json:"type"`
	Snapshots []HealthSnapshot `json:"snapshots"`
	Count     int              `json:"count"`
	Timestamp time.Time        `json:"timestamp"`
}
